plugins/funind/functional_principles_proofs.ml

```diff
@@ -861,8 +861,10 @@ let generate_equation_lemma evd fnames f fun_num nb_params nb_args rec_args_num
   let lemma, _ =
     Declare.Proof.by (Proofview.V82.tactic prove_replacement) lemma
   in
-  let (_ : _ list) =
-    Declare.Proof.save ~proof:lemma ~opaque:Vernacexpr.Transparent ~idopt:None
+  let pm = Declare.OblState.empty in
+  let _pm, _ =
```

Comment:
This is not very nice, we shouldn't have to care about program stuff when we don't use it.

Reply:
Indeed it is a bit sketchy, you could see this from two points of view:

- funind uses the API in a kind of weird way [in fact it is the only part of the code that needs this], we should eventually fix funind but not a big priority.
- we should classify proofs in a finer way, so `Proof.start` doesn't indeed care about obligations; this is planned, but ATM is not immediate as `save` has to work for both obligations / regular lemmas [see declare and the match on `Proof_info.t`]

What to do here is still a bit open IMO.

Reply:
> What to do here is still a bit open IMO.

In particular, it would interesting to think whether we want obligations state to be part of the proof state, thus the evar_map of the original term with holes persists among commands; this would be interesting and should avoid quite a few hacks, but would imply that closing a proof keeps some proof state open.

That's perfectly manageable in with the vernac interpretation infrastructure, by having `VtCloseProof` return a (reified) continuation; I'm not sure if the STM would boom about it tho.

Reply:
Why not making pm an option and pass :None almost everywhere? (to signal we are not using obligations, while obl.empty looks like the initial state of a process involving some obligations).

Reply:
Actually the pm parameters is used everywhere accurately except in funind, so I'm not sure if funding should be instead fixed.

As I mentioned above I'd be cool to have a bit of better typing , but I'm unsure about the way to go, qed can be used in an obligation, so we could have another version of save only for funind, but I dunno, still I don't see the big picture here, so this kind of pr is mostly about making explicit what is already going on. Suggestions / ideas very welcome!

Reply:
This still bugs me so I added a commit to deal with it, please review.